### PR TITLE
WIP: Fix missing namespace resource when backing up namespaces with auto cluster resources

### PR DESCRIPTION
**Pull request by dgoodwin**

### CODE REVIEW

**Comment by ncdc** [comment]
Not quite @nrb. @dgoodwin is working through one issue.

File: pkg/backup/resource_backupper.go [hunk 1]
```diff
@@ -174,7 +176,7 @@ func (rb *defaultResourceBackupper) backupResource(
 	)
 
 	var namespacesToList []string
-	if resource.Namespaced {
+	if resource.Namespaced || resource.Kind == "Namespace" {
```
Comment (ncdc) [inline pkg/backup/resource_backupper.go]: As we discussed on Slack, this doesn't do what you think it might. If you call `rb.dynamicFactory.ClientForGroupVersionResource(gv, resource, namespace)` where `resource` is the namespaces resource, and `namespace` is a specific name of a namespace, the call that immediately follows:

```go
unstructuredList, err := resourceClient.List(metav1.ListOptions{LabelSelector: rb.labelSelector})
```

will still return a full listing of all of the namespaces.

**Comment by dgoodwin** [comment]
Ok updated with passing unit tests, and manual testing, this does fix the issue I hit.

File: pkg/backup/resource_backupper.go [hunk 2]
```diff
@@ -174,37 +176,57 @@ func (rb *defaultResourceBackupper) backupResource(
...
-	if resource.Namespaced {
-		namespacesToList = getNamespacesToList(rb.namespaces)
-	} else {
-		namespacesToList = []string{""}
-	}
-	for _, namespace := range namespacesToList {
-		resourceClient, err := rb.dynamicFactory.ClientForGroupVersionResource(gv, resource, namespace)
+	// If processing the Namespace resource specifically and you've specified to include only some,
+	// fetch them individually instead of a full list of all in the cluster:
+	if resource.Kind == "Namespace" && !rb.namespaces.IncludeEverything() {
```
Comment (ncdc) [inline pkg/backup/resource_backupper.go]: The `!rb.namespaces.IncludeEverything()` check isn't going to work here. This expression will evaluate to true if you have any excludes, even if you don't have any specific includes, for example. You're better off checking if length of `GetIncludes()` >= 1 and it doesn't contain `*`. Or we could create another helper `HasSpecificIncludes` or something like that.

Comment (dgoodwin) [reply pkg/backup/resource_backupper.go]: Updated with another pass at that, I see validation that if * is in the list it must be the only element in the list, so I hope I got it right this time.

File: pkg/backup/resource_backupper.go [hunk 3]
```diff
@@ -174,37 +176,59 @@ func (rb *defaultResourceBackupper) backupResource(
...
-	if resource.Namespaced {
-		namespacesToList = getNamespacesToList(rb.namespaces)
-	} else {
-		namespacesToList = []string{""}
-	}
-	for _, namespace := range namespacesToList {
-		resourceClient, err := rb.dynamicFactory.ClientForGroupVersionResource(gv, resource, namespace)
+	// If processing the Namespace resource specifically and you've specified to include only some,
+	// fetch them individually instead of a full list of all in the cluster:
+	if resource.Kind == "Namespace" &&
```
Comment (ncdc) [inline pkg/backup/resource_backupper.go]: Probably should also make sure that gv.Group == ""

File: pkg/backup/resource_backupper.go [hunk 4]
```diff
@@ -174,37 +176,59 @@ func (rb *defaultResourceBackupper) backupResource(
...
+		!(len(rb.namespaces.GetIncludes()) == 0 || rb.namespaces.GetIncludes()[0] == "*") {
```
Comment (ncdc) [inline pkg/backup/resource_backupper.go]: This could be written in the positive as `(len(rb.namespaces.GetIncludes()) >= 1 && rb.namespaces.GetIncludes()[0] != "*")`. Maybe a bit easier to read?

File: pkg/backup/resource_backupper.go [hunk 5]
```diff
@@ -174,37 +176,59 @@ func (rb *defaultResourceBackupper) backupResource(
...
+
+		resourceClient, err := rb.dynamicFactory.ClientForGroupVersionResource(gv, resource, "")
 		if err != nil {
 			return err
 		}
+		for _, namespace := range rb.namespaces.GetIncludes() {
+			// TODO: labelSelector relevant here?
```
Comment (ncdc) [inline pkg/backup/resource_backupper.go]: It is relevant, but GetOptions doesn't support it. You'd have to get a selector object and call Matches on it by hand.

Maybe we should just ignore this optimization for the time being?

**Comment by dgoodwin** [comment]
Updated to skip the optimization for now as it looked more complicated than expected, left a todo. Looking ok so far in my testing.

**Comment by ncdc** [comment]
LGTM. WDYT @skriss? I'd probably eventually like a unit test that doesn't mock out the itemBackupper completely so we can make sure that when you are only including certain namespaces, the itemBackupper filters out a namespace that doesn't match.

**Comment by dgoodwin** [comment]
You may already have such a test: https://github.com/heptio/ark/blob/master/pkg/backup/item_backupper_test.go#L51

**Comment by ncdc** [comment]
Yes, but I want one that ensures that when the resourceBackupper is listing namespaces and sending them one by one to the itemBackupper, the itemBackupper filters out ones that are in the includes list.

**Comment by skriss** [comment]
I'm probably missing something, but isn't this going to end up backing up all namespaces? since namespaces themselves are cluster-scoped, itemBackupper isn't going to filter them out based on namespace includes/excludes..

File: pkg/backup/resource_backupper.go [hunk 6]
```diff
@@ -129,7 +129,9 @@ func (rb *defaultResourceBackupper) backupResource(
 		// (all namespaces) backup. Note that in the case of a subset of
 		// namespaces being backed up, some related cluster-scoped resources
 		// may still be backed up if triggered by a custom action (e.g. PVC->PV).
-		if !resource.Namespaced && !rb.namespaces.IncludeEverything() {
+		// If we're processing namespaces themselves, we will not skip here, they may be
+		// filtered out later.
+		if !resource.Namespaced && resource.Kind != "Namespace" && !rb.namespaces.IncludeEverything() {
```
Comment (nrb) [inline pkg/backup/resource_backupper.go]: This will skip if the resource is a namespace, but doesn't appear to filter based on included/excluded namespaces. Is that intentional?

**Comment by ncdc** [comment]
@skriss you're right. Back to the drawing board.

**Comment by skriss** [comment]
@ncdc see my updated comment ^^^ - does that seem reasonable?

TLDR namespace API objects are kind of a pain.

File: pkg/backup/resource_backupper.go (hunk 6 above)
Comment (ncdc) [reply pkg/backup/resource_backupper.go]: It was, until we realized that the itemBackupper's filtering of namespaces based on includes/excludes only applies to namespace-scoped resources, but it won't work for namespaces (they're cluster-scoped).

**Comment by ncdc** [comment]
Could work

File: pkg/backup/resource_backupper.go (hunk 6 above)
Comment (nrb) [reply pkg/backup/resource_backupper.go]: Right, I saw that code earlier. Is there some sort of cluster-scoped filtering that could be introduced?

Comment (ncdc) [reply pkg/backup/resource_backupper.go]: This is only messy because it's the namespace object itself

**Comment by ncdc** [comment]
Replaced by #182
